google/cloud/spanner/internal/time.cc

```diff
@@ -0,0 +1,216 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/time.h"
+#include <chrono>
+#include <cstdint>
+#include <cstring>
+#include <ctime>
+#include <iomanip>
+#include <ratio>
+#include <sstream>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+// All the civil-time code assumes the proleptic Gregorian calendar, and
+// 24-hour days divided into 60-minute hours and 60-second minutes.
+
+using nanoseconds = std::chrono::nanoseconds;
+using seconds = std::chrono::seconds;
+using system_clock = std::chrono::system_clock;
+using time_point = system_clock::time_point;
+
+//
+// Duration
+//
+
+google::protobuf::Duration ToProto(nanoseconds ns) {
+  auto nanos = ns.count();
+  google::protobuf::Duration proto;
+  proto.set_seconds(nanos / nanoseconds::period::den);  // rounds toward zero
+  proto.set_nanos(nanos % nanoseconds::period::den);    // (a/b)*b + a%b == a
+  return proto;
+}
+
+void ToProto(nanoseconds ns, google::protobuf::Duration* proto) {
+  *proto = ToProto(ns);
+}
+
+nanoseconds FromProto(google::protobuf::Duration const& proto) {
+  nanoseconds ns(proto.nanos());
+  ns += seconds(proto.seconds());
+  return ns;
+}
+
+//
+// Timestamp
+//
+
+google::protobuf::Timestamp ToProto(time_point tp) {
+  std::time_t t = system_clock::to_time_t(tp);
+  time_point ttp = system_clock::from_time_t(t);
+  auto ss = std::chrono::duration_cast<nanoseconds>(tp - ttp);
+  if (ss.count() < 0) {
+    t -= 1;
+    ss += seconds(1);
+  }
+  google::protobuf::Timestamp proto;
+  proto.set_seconds(t);
+  proto.set_nanos(ss.count());
+  return proto;
+}
+
+void ToProto(time_point tp, google::protobuf::Timestamp* proto) {
+  *proto = ToProto(tp);
+}
+
+time_point FromProto(google::protobuf::Timestamp const& proto) {
+  time_point tp = system_clock::from_time_t(proto.seconds());
+  tp += nanoseconds(proto.nanos());
+  return tp;
+}
+
+namespace {
+
+// A duration capable of holding subsecond values at high precision.
+using femtoseconds = std::chrono::duration<std::int64_t, std::femto>;
+
+// Convert a std::time_t into a Zulu std::tm.
+std::tm ZTime(std::time_t const t) {
+  std::time_t sec = t % (24 * 60 * 60);
+  std::time_t day = t / (24 * 60 * 60);
+  if (sec < 0) {
+    sec += 24 * 60 * 60;
+    day -= 1;
+  }
+  day += 719468;  // since 1970-01-01
```

Comment:
Since 1970-01-01 to when?  I guess 1900 as `std::tm` is 1900 based?

Reply:
719468 is the difference in days between 1970-01-01 and 0000-03-01.  I added a direct comment about that (and the use of Mar 1st instead of Jan 1st).  The 1900 for `tm_year` appears only in its immediate reads/writes.